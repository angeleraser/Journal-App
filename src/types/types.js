export const TYPES = {
  login: "[Auth] Login",
  logout: "[Auth] Logout",

  uiSetError: "[UI] Set Error",
  uiRemoveError: "[UI] Remove Error",

  uiStartLoading: "[UI] Start Loading",
  uiFinishLoading: "[UI] Finish Loading",

  notesAddNew: "[Notes] Add new note",
  notesSelect: "[Notes] Select note",
  notesLoad: "[Notes] Load notes",
  notesUpdate: "[Notes] Update note",
  notesFileURL: "[Notes] Update image url",
  notesDelete: "[Notes] Delete note",
  notesLogoutCleaning: "[Notes] Logout cleaning",
};
