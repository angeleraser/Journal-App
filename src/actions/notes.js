import Swal from "sweetalert2";
import { dataBase } from "../firebase/firebase-config";
import { uploadFile } from "../helpers/fileUpload";
import { fetchUserInfoFromDatabase } from "../helpers/loadNotes";
import { TYPES } from "../types/types";

export const addNewNote = () => {
  return async (dispatch, getState) => {
    // getState returns the current main state and it can be used in async operations
    const {
      auth: { uID },
    } = getState();
    const newEntry = {
      body: "",
      title: "",
      date: new Date().getTime(),
      imgURL: null,
    };
    /*
     Collection journal by user ID 
    */
    const DOCUMENT_REF = await dataBase
      .collection(`${uID}/journal/notes`)
      .add(newEntry);
    dispatch(selectNote(DOCUMENT_REF.id, newEntry));
    dispatch(updateNoteList(DOCUMENT_REF.id, newEntry));
  };
};

export const selectNote = (id, note) => {
  return {
    type: TYPES.notesSelect,
    payload: {
      id,
      ...note,
    },
  };
};

export const setNotes = (notes) => {
  return {
    type: TYPES.notesLoad,
    payload: notes,
  };
};

export const startLoadingNotes = (uID) => {
  return async (dispatch) => {
    const notes = await fetchUserInfoFromDatabase(uID, "journal/notes");
    dispatch(setNotes(notes));
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {
      auth: { uID },
    } = getState();
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    await dataBase
      .doc(`${uID}/journal/notes/${note.id}`)
      .update(noteToFirestore);
    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: TYPES.notesUpdate,
  payload: {
    id,
    note,
  },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    const { selected } = getState().notes;
    Swal.fire({
      title: "Uploading...",
      text: "Please wait a few seconds...",
      allowOutsideClick: false,
      onBeforeOpen() {
        Swal.showLoading();
      },
    });
    const fileURL = await uploadFile(file);
    selected.imgURL = fileURL;
    Swal.close();
    dispatch(startSaveNote(selected));
  };
};

export const startDeletingNote = (id) => {
  return async (dispatch, getState) => {
    const uID = getState().auth.uID;
    dataBase.doc(`${uID}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: TYPES.notesDelete,
  payload: {
    id,
  },
});

export const cleanNotesAfterLogout = () => ({
  type: TYPES.notesLogoutCleaning,
});

export const updateNoteList = (id, note) => {
  return {
    type: TYPES.notesAddNew,
    payload: {
      id,
      ...note,
    },
  };
};
