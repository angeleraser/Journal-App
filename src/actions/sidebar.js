import { TYPES } from "../types/types";

export const openSidebar = () => ({
  type: TYPES.sidebarOpen,
});

export const closedSidebar = () => ({
  type: TYPES.sidebarClosed,
});
