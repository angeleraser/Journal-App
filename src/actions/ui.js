import { TYPES } from "../types/types";

export const setUiError = (errorMsg) => {
  return {
    type: TYPES.uiSetError,
    payload: errorMsg,
  };
};

export const removeUiError = () => {
  return {
    type: TYPES.uiRemoveError,
  };
};
