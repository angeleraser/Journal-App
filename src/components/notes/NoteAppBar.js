import React from "react";

export const NoteAppBar = () => {
  return (
    <div className="notes__appbar w-full flex items-center justify-between">
      <h2>28 de agosto 2020</h2>
      <div className='flex items-center'>
        <button className="btn btn-rounded btn-shadow">Picture</button>
        <button className="btn btn-rounded btn-shadow">Save</button>
      </div>
    </div>
  );
};
