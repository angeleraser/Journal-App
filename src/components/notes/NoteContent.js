import React from 'react'

export const NoteContent = () => {
  return (
    <div>
      
    </div>
  )
}
