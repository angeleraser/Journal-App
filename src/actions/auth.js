import Swal from "sweetalert2";
import { FIREBASE, googleAuthProvider } from "..";
import { TYPES } from "../types/types";
import { uiStartLoading, uiFinishLoading } from "./ui";
// NOTA: Si la tarea es asincrona la funcion debe retonar otra funcion que realiza el dispatch, sino un objeto con el tipo de accion y el payload normal.

export const login = (uID, displayName) => ({
  type: TYPES.login,
  payload: {
    uID,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithEmailAndPassword(
          email,
          password
        );
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().createUserWithEmailAndPassword(
          email,
          password
        );
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithPopup(
          googleAuthProvider
        );
        dispatch(login(user.id, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await FIREBASE.auth().signOut();
    dispatch(logout());
  };
};

export const logout = () => ({
  type: TYPES.logout,
});
