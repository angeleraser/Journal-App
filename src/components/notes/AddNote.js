/* eslint-disable react-hooks/exhaustive-deps */
import React from "react";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { selectNote, startDeletingNote } from "../../actions/notes";
import useForm from "../hooks/useForm";

export const AddNote = () => {
  const { selected } = useSelector(({ notes }) => notes);
  const dispatch = useDispatch();
  const [formValues, handleInputChange, reset] = useForm({ ...selected });
  const { body, title } = formValues;
  useEffect(() => {
    reset(selected);
  }, [selected]);

  // Update the current note when the user edit the note
  useEffect(() => {
    dispatch(selectNote(selected.id, { ...formValues }));
  }, [body, title]);

  const handleDeleteNote = () => {
    dispatch(startDeletingNote(selected.id));
  };
  return (
    <div className="flex flex-col notes__add-note">
      <input
        name="title"
        onChange={handleInputChange}
        type="text"
        value={title}
        placeholder="Add a title"
      />
      <textarea
        onChange={handleInputChange}
        name="body"
        value={body}
        placeholder="What happened today?"></textarea>
      {!!selected.imgURL && (
        <div className="note-image">
          <img className="h-full btn-shadow" src={selected.imgURL} alt="note" />
        </div>
      )}
      <button onClick={handleDeleteNote} className="btn btn-danger">
        Delete
      </button>
    </div>
  );
};
