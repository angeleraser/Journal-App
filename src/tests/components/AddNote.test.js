import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { AddNote } from "../../components/notes/AddNote";
import { act } from "react-dom/test-utils";
import { selectNote } from "../../actions/notes";
import { TYPES } from "../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Hay que definir el estado inicial igual al de la aplicacion en ese punto
const initState = {
  auth: {
    name: "angeleraser",
    uID: "k1j2g5bk75j23",
    photoURL: null,
  },
  notes: {
    notes: [],
    selected: {
      id: 1234,
      title: "Mi nota",
      body: "body",
      imgURL: "https://sadasdas.com",
    },
  },
  ui: {
    loading: false,
    errorMsg: null,
  },
  sidebar: {
    isOpen: false,
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <AddNote />
    </MemoryRouter>
  </Provider>
);
// Fix dispatch del store si las acciones a evaluar son asincronas
jest.mock("../../actions/notes", () => {
  return { selectNote: jest.fn() };
});
describe("Testing <Sidebar/>", () => {
  // 1-
  test("should match with snap", () => {
    expect(wrapper).toMatchSnapshot();
  });
  //  2-
  test("should dispatch selectNote action", () => {
    const textarea = wrapper.find("input[type='text']");
    expect(textarea.exists()).toBe(true);
    textarea.simulate("change", {
      target: {
        name: "title",
        value: "New body",
      },
    });
    expect(selectNote).toHaveBeenLastCalledWith(initState.notes.selected.id, {
      title: "New body",
      body: "body",
      id: 1234,
      imgURL: "https://sadasdas.com",
    });
  });
});
