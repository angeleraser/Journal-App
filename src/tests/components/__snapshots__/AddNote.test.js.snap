// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Sidebar/> should match with snap 1`] = `
<div
  className="flex flex-col notes__add-note"
>
  <input
    autoComplete="off"
    name="title"
    onChange={[Function]}
    placeholder="Add a title"
    type="text"
    value="Mi nota"
  />
  <textarea
    autoComplete="off"
    name="body"
    onChange={[Function]}
    placeholder="What happened today?"
    value="body"
  />
  <div
    className="note-image"
  >
    <img
      alt="note"
      className="h-full btn-shadow"
      src="https://sadasdas.com"
    />
  </div>
  <button
    className="btn delete-btn"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
